### PR TITLE
[IMP] web_editor, website: add options to upload documents on links

### PR DESCRIPTION
**Pull request by younn-o**
It is now possible to upload a file via the link tools, and link it to a
button.

An option was added when a link is used for serving a document to
specify how it should be served : either opening it in the browser or
downloading it.

As documents are handled through links, they won't be selectable via the
media dialog opened with "Insert media" or double clicking on an image.

task-2172311

Co-authored-by: Matthieu Stockbauer <tsm@odoo.com>

Spec validated are https://github.com/odoo/odoo/pull/69519#issuecomment-826977005

### CODE REVIEW

File: addons/web_editor/static/src/js/wysiwyg/widgets/link.js [hunk 1]
```diff
@@ -234,10 +234,10 @@ const Link = Widget.extend({
      * @private
      */
     _correctLink: function (url) {
-        if (url.indexOf('mailto:') === 0 || url.indexOf('tel:') === 0) {
-            url = url.replace(/^tel:([0-9]+)$/, 'tel://$1');
-        } else if (url.indexOf('@') !== -1 && url.indexOf(':') === -1) {
+        if (url.indexOf('@') >= 0 && url.indexOf('mailto:') === -1 && !url.match(/^http[s]?/i)) {
```
Comment (rdeodoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link.js:237]:
```suggestion
        if (url.includes('@') && !url.includes('mailto:')  && !url.match(/^http[s]?/i)) {
```
Why not use includes? Feels easier to read

File: addons/web_editor/static/src/js/wysiwyg/widgets/link.js [hunk 2]
```diff
@@ -234,10 +234,10 @@ const Link = Widget.extend({
...
             url = 'mailto:' + url;
+        } else if (/^[^a-zA-Z]*[0-9]+[^a-z]*$/.test(url)) {
```
Comment (rdeodoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link.js:239]: Why `[^a-z]` and not `[^a-zA-Z]` ?
Is this for the american phone number with letters?
If thats the case:
1. I give you 10 bonus point for taking that into consideration
2. I remove 1 bonus point for not adding a comment about it (no one will ever know).

Note that it will recognize `01NET.COM` as a phone but I guess we can live with that

Comment (rdeodoo) [reply addons/web_editor/static/src/js/wysiwyg/widgets/link.js]: No extra point then :D

Comment (younn-o) [reply addons/web_editor/static/src/js/wysiwyg/widgets/link.js]: Nope... that was just a mistake :grimacing:

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/wysiwyg/widgets/link.js]: Where can I spend RDE points? :open_mouth:

**Comment by rdeodoo** [comment]
For tracability, this comes from #69519

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Commit about phone number LGTM
Comments here are about the other commit.

Also, I am not convinced by the fact write the full URL of the file as text when creating a link
See here, I uploaded `Order - S00027.pdf`, it would be nicer to have that as link than
image


Note: If not co-authored, I'd at least add `Courtesy of TSM` inside your commit message as it is quite largely inspired by https://github.com/odoo/odoo/pull/69519/files.

File: addons/web_editor/static/src/xml/wysiwyg.xml [hunk 3]
```diff
@@ -510,7 +510,45 @@
                 </div>
             </we-input>
         </we-row>
-        <we-row class="o_strip_domain d-none" t-attf-class="#{widget.isButton ? ' d-none' : ''}">
+        <we-row>
+            <we-title>⌙ Media</we-title>
+            <div t-attf-class="{{widget.fileName ? '' : 'd-none'}} o_we_edit_file_wrapper">
+                <span class="text-truncate o_we_filename"><t t-out="widget.fileName"/></span>
+                <we-button class="o_we_bg_success o_we_upload_file" style="min-width: 29.62px;" title="Replace file">
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/web_editor/static/src/xml/wysiwyg.xml]: Indeed, will be fixed with https://github.com/odoo/odoo/pull/73241/files#diff-f1acdd04893b739b62e26ae33945d012c4ce8df627f23ea84968f707369c8eaeR842
You don't have to care about that, we can merge it with 
image

**Comment by rdeodoo** [comment]
Should target 14.5 IMO, nice to have for 15.0 and no model/view change so can go in stable

File: addons/web_editor/static/src/xml/wysiwyg.xml [hunk 4]
```diff
@@ -509,7 +509,42 @@
...
+        <we-row>
+            <we-title>⌙ Media</we-title>
+            <div class="o_we_edit_file_wrapper">
+                <span class="text-truncate o_we_filename"><t t-out="widget.fileName"/></span>
+                <we-button class="o_we_bg_success o_we_upload_file" title="Replace file">
+                    <i class="fa fa-fw fa-refresh"/>
+                </we-button>
+                <we-button class="o_we_bg_danger o_we_remove_file" title="Remove file">
+                    <i class="fa fa-fw fa-trash"/>
+                </we-button>
```
Comment (rdeodoo) [inline addons/web_editor/static/src/xml/wysiwyg.xml:521]:
```suggestion
                <we-button class="o_we_button_icon o_we_bg_success o_we_upload_file fa fa-fw fa-refresh" title="Replace file"/>
                <we-button class="o_we_button_icon o_we_bg_danger o_we_remove_file fa fa-fw fa-trash" title="Remove file"/>
```
with https://github.com/odoo/odoo/pull/73241/commits/de51b83cdb369db17029befce03c4cf3e14a57f7#r710958930 ofc

File: addons/web_editor/static/src/xml/wysiwyg.xml [hunk 5]
```diff
@@ -509,7 +509,42 @@
...
+                            Download
+                        </we-button>
+                        <we-button t-attf-class="dropdown-item" href="#" data-value="">
+                            Open
+                        </we-button>
+                    </we-selection-items>
+                </div>
+            </we-select>
+        </we-row>
+        <we-row class="o_strip_domain d-none" t-attf-class="o_strip_domain #{widget.isButton ? ' d-none' : ''}">
```
Comment (rdeodoo) [inline addons/web_editor/static/src/xml/wysiwyg.xml:547]: Why adding `o_strip_domain` on t-attf-class?
One or the other should go, but you can't have both.
In qweb python, the t-attf-class will only be taken into consideration
In qweb js I wasnt sure but I just tested and it seems like the class will be used (so the opposite)
So you diff on this line actually does nothing if I am not wrong

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js [hunk 6]
```diff
@@ -320,6 +383,37 @@ const LinkTools = Link.extend({
...
+     * @private
+     * @param {Event} ev
+     */
+    _onUploadFileClick(ev) {
+        const documentDialog = new MediaDialog(this, {noIcons: true, noVideos: true, noImages: true});
+        documentDialog.on('save', this, data => {
+            if (data) {
+                const {src, href, title} = data;
+                let url = href || src;
+                url = url.replace(/[&?]+download=true/, '').replace(window.location.origin, '');
```
Comment (rdeodoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js:396]: I'd add a comment here, you can't resolve this comment if you think this is not needed/is obvious without the comment:
something like
// download=true automatically added by dialog save in js, we don't want it, default file link should open in need tab (and will be configurable in link tool)

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Missing the target _blank on links of type "open"
After that this should be LGTM

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js (hunk 6 above)
Comment (younn-o) [reply addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js:399]: done

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Found 2 bugs while testing
1. Click on an existing link (eg Home on footer), the anchor option will be shown, now add a file, the anchor option remain visible
2. Create a link with a file, remove the file, the target=_blank remain on the link, even if the link is now /contactus eg

**Comment by younn-o** [comment]
> Found 2 bugs while testing
> 
> 1. Click on an existing link (eg Home on footer), the anchor option will be shown, now add a file, the anchor option remain visible
> 2. Create a link with a file, remove the file, the target=_blank remain on the link, even if the link is now /contactus eg

@rdeodoo Done

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js [hunk 7]
```diff
@@ -65,6 +70,25 @@ const LinkTools = Link.extend({
...
+        }
+        this._super(...arguments);
+
+        if (this.fileName) {
+            if (this._getFileAction() === 'download') {
+                this.linkEl.setAttribute('download', '');
+            } else {
+                this.linkEl.removeAttribute('download');
+            }
+        }
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js:91]: Should not the attribute also be removed if `!this.fileName` ?

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js [hunk 8]
```diff
@@ -116,6 +141,15 @@ const LinkTools = Link.extend({
     _getLinkType: function () {
         return this.$('we-selection-items[name="link_style_color"] we-button.active').data('value') || '';
     },
+    /**
+     * Returns the opening methode for the file
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js:146]: - method?
- Point at the end of doc sentences

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js [hunk 9]
```diff
@@ -212,17 +246,45 @@ const LinkTools = Link.extend({
...
         }
+        this.el.querySelector('.o_we_upload_file_wrapper').classList.toggle('d-none', !!this.fileName);
+        this.el.querySelector('.o_we_edit_file_wrapper').classList.toggle('d-none', !this.fileName);
+        this.el.querySelector('.o_we_file_action_wrapper').classList.toggle('d-none', !this.fileName);
+        this.el.querySelector('.o_we_filename').textContent = this.fileName;
+        if (this.fileName) {
+            this.el.querySelector('input[name="url"]').setAttribute('readonly', '');
+        } else {
+            this.el.querySelector('input[name="url"]').removeAttribute('readonly');
+        }
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js:258]: You can use the `readonly` property for this:
```javascript
this.el.querySelector('input[name="url"]').readonly = !!this.fileName;
```

File: addons/web_editor/static/src/scss/wysiwyg_snippets.scss [hunk 10]
```diff
@@ -883,6 +883,11 @@ body.editor_enable.editor_has_snippets {
             }
         }
 
+        // Buttons with an icon but no text, so no possible overflow
+        we-button.o_we_button_icon {
+            flex: 0 0 auto;
+        }
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/scss/wysiwyg_snippets.scss:889]: There should be a dedicated commit about this... or is it the first button with an icon that we introduce in the panel? :thinking: I doubt it (and if not the first button, those buttons should be fixed, probably in stable versions). As I explained to @rdeodoo I also would like the developer/designer to not have to think about adding this class:
- Either it's a button with a fa icon: it should be define with `<we-button class="... fa fa-machin ..."` -> the CSS rule can detect `we-button.fa` (as I thought it was already the case at some point)
- Or it's a button with a image icon -> it is defined with `<we-button data-img="..."` -> the JS which sets up that image can also automatically add the `o_we_button_icon` class if necessary

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js [hunk 11]
```diff
@@ -49,8 +49,9 @@ const LinkTools = Link.extend({
         this.$link.addClass('oe_edited_link');
         this.$button.addClass('active');
         return this._super(...arguments).then(() => {
-            if (!this.options.noFocusUrl) {
-                dom.scrollTo(this.$(':visible:last')[0], {duration: 0});
+            const lastOptionEl = this.$(':visible:last')[0];
+            if (!this.options.noFocusUrl && lastOptionEl) {
+                dom.scrollTo(lastOptionEl, {duration: 0});
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js:54]: Should have a comment to explain how we can reach this code and not have a visible element, I am not sure to understand. Also, it should be for 14.4 I guess?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/wysiwyg/widgets/link.js [hunk 12]
```diff
@@ -88,6 +90,10 @@ const Link = Widget.extend({
             this.data.content = this.data.content ? this.data.content.replace(/[ \t\r\n]+/g, ' ') : '';
         }
 
+        if (this.linkEl) {
+            this.data.isNewWindow = this.data.isNewWindow || this.linkEl.target === '_blank';
+        }
```
[the remark this answers is not in the document]
Comment (younn-o) [reply addons/web_editor/static/src/js/wysiwyg/widgets/link.js:95]: I opened a new PR for saas-14.4 here https://github.com/odoo/odoo/pull/76846

File: addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js (hunk 11 above)
Comment (younn-o) [reply addons/web_editor/static/src/js/wysiwyg/widgets/link_tools.js]: I have not been able to reproduce it on 14.4. It seems to be related to the _onMouseDown method of the editor which differs between the two versions. I proposed another fix on this branch

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 13]
```diff
@@ -3187,7 +3187,11 @@ var SnippetsMenu = Widget.extend({
             $(range.commonAncestorContainer).parent('[data-oe-model]:not([data-oe-type="html"]):not([data-oe-field="arch"])').length ||
             (e && $(e.target).closest('.fa, img').length ||
             this.options.wysiwyg.lastMediaClicked && $(this.options.wysiwyg.lastMediaClicked).is('.fa, img')) ||
-            (this.options.wysiwyg.lastElement && !this.options.wysiwyg.lastElement.isContentEditable)
+            // An additional test to check if the last element is a link on top
+            // of being editable is mandatory as the editor temporarily makes
+            // them not editable on mouse down.
+            (this.options.wysiwyg.lastElement && this.options.wysiwyg.lastElement.tagName !== 'A'
+                && !this.options.wysiwyg.lastElement.isContentEditable)
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:3194]: The whole original condition seems ultra messy :grimacing: but well... Anyways, not sure to understand this change as I don't understand the flow you describe in the commit message, can you re-explain or show me ? ;) This is probably better than the other fix you made though, however the other fix fixed something else: 
Screenshot from 2021-09-20 18-08-32
-> Trying to edit a really non-editable link -> the popover still is shown (ok) -> but the edit button is shown and clicking on it crashes (in this case, the fix should be to hide the button or at worst the popover I guess... cc @rdeodoo

Comment (rdeodoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:3194]: That's some links we never detected and that should be flagged as non editable. I am afraid there is no discriminant for this (kind of) link(s), so we should just add `o_no_link_popover` on those links from the template.

Comment (younn-o) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:3194]: After investigating a bit more, the bug happens when clicking on a link that is displayed behind the edit_menu popover. 
Here is a video

https://user-images.githubusercontent.com/58561718/134140565-69d9d3ae-a239-4883-a256-f3d21f60c878.mp4

In the end I am not sure it is directly linked to the editor setting the link contenteditable attribute to false then to true to correctly display the cursor... 
Reverting the commit 4eff35ab31bfc45b4c8dca8392e9da4f4e2a7afd removes the bug, so it may be related to using "focus" as a trigger for the popover... I am still trying to understand what bootstrap is doing in that case

File: addons/web_editor/static/src/scss/wysiwyg_snippets.scss [hunk 14]
```diff
@@ -884,8 +884,9 @@ body.editor_enable.editor_has_snippets {
         }
 
         // Buttons with an icon but no text, so no possible overflow
-        we-button.o_we_button_icon {
+        we-button.fa {
```
Comment (rdeodoo) [inline addons/web_editor/static/src/scss/wysiwyg_snippets.scss:887]: @younn-o Can you extract this to a separate PR?
It seems like multiple PR need it. So we can merge it quickly.

For the reminder, it was added with https://github.com/odoo/odoo/pull/73241/files#diff-f1acdd04893b739b62e26ae33945d012c4ce8df627f23ea84968f707369c8eaeR887
Being improved here (automatic instead of class required).
But also needed for #74206

Comment (younn-o) [reply addons/web_editor/static/src/scss/wysiwyg_snippets.scss]: I extracted it here : https://github.com/odoo/odoo/pull/77173

File: addons/web_editor/static/src/js/editor/snippets.editor.js (hunk 13 above)
Comment (rdeodoo) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: The changes about the popover (manual trigger) LGTM

File: addons/web_editor/static/src/js/wysiwyg/widgets/link.js [hunk 15]
```diff
@@ -257,13 +257,15 @@ const Link = Widget.extend({
...
-                    && url[0] !== '#' && url.slice(0, 2) !== '${') {
-            url = 'http://' + url;
+        if (!url.match(/^http[s]?/i)) {
+            if (url.includes('@') && !url.startsWith('mailto:')) {
+                url = 'mailto:' + url;
+            } else if (/^[^a-zA-Z]*[0-9]+[^a-zA-Z]*$/.test(url) && !url.startsWith('tel:')) {
+                url = 'tel:' + url;
+            } else if (!url.includes('://') && url[0] !== '/'
+                        && url[0] !== '#' && url.slice(0, 2) !== '${') {
+                url = 'http://' + url;
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/link.js:267]: How can I receive an URL which start by http but does not include `://` (and in which case do I want to add http at the start when there is already http at the start)... seems really wrong.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
If it was perfect, I would have merged now but it's not. Need to review everything in pair at some point. @younn-o I guess it will be easier to wait for the merge of frontend to backend and adapt this one afterwards. I rebased and retargeted to master. I also reviewed the first commit entirely. Added some comments running over the rest but really did not check in details.

EDIT: oops forget the second comment, wanted to remove it but forgot before sending the review ;)

**Comment by C3POdoo** [comment]
Dear @younn-o,

Thank you for your contribution but we are closing it due to inactivity.

We apology if we could not look at your request in time.
If the contribution still makes sense, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

_This is an automated message._

**Comment by qsm-odoo** [comment]
Time to close this one. Main purpose was merged with https://github.com/odoo/odoo/commit/cec6ee74de4b904b590d41140d99ae844ed1f15d, other purposes can be revived later if needed :shrug:
